test/adaptive_load/scoring_function_test.cc

```diff
@@ -0,0 +1,231 @@
+  config.mutable_lower_threshold()->set_value(5.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(4.0), -1.0);
+}
+
+TEST(BinaryScoringFunctionTest, ReturnsOneForValueWithinThresholdRange) {
+  nighthawk::adaptive_load::BinaryScoringFunctionConfig config;
+  config.mutable_lower_threshold()->set_value(5.0);
+  config.mutable_upper_threshold()->set_value(7.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(6.0), 1.0);
+}
+
+TEST(BinaryScoringFunctionTest, ReturnsOneForValueEqualToLowerThresholdOfRange) {
+  nighthawk::adaptive_load::BinaryScoringFunctionConfig config;
+  config.mutable_lower_threshold()->set_value(5.0);
+  config.mutable_upper_threshold()->set_value(7.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(5.0), 1.0);
+}
+
+TEST(BinaryScoringFunctionTest, ReturnsOneForValueEqualToUpperThresholdOfRange) {
+  nighthawk::adaptive_load::BinaryScoringFunctionConfig config;
+  config.mutable_lower_threshold()->set_value(5.0);
+  config.mutable_upper_threshold()->set_value(7.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(7.0), 1.0);
+}
+
+TEST(BinaryScoringFunctionTest, ReturnsNegativeOneForValueBelowThresholdRange) {
+  nighthawk::adaptive_load::BinaryScoringFunctionConfig config;
+  config.mutable_lower_threshold()->set_value(5.0);
+  config.mutable_upper_threshold()->set_value(7.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(4.0), -1.0);
+}
+
+TEST(BinaryScoringFunctionTest, ReturnsNegativeOneForValueAboveThresholdRange) {
+  nighthawk::adaptive_load::BinaryScoringFunctionConfig config;
+  config.mutable_lower_threshold()->set_value(5.0);
+  config.mutable_upper_threshold()->set_value(7.0);
+  BinaryScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(8.0), -1.0);
+}
+
+TEST(LinearScoringFunctionTest, ReturnsZeroForValueEqualToThreshold) {
+  nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+  config.set_threshold(10.0);
+  config.set_k(1.0);
+  LinearScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(10.0), 0.0);
+}
+
+TEST(LinearScoringFunctionTest, ReturnsPositiveValueForValueBelowThreshold) {
+  nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+  config.set_threshold(10.0);
+  config.set_k(1.0);
+  LinearScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(8.0), 2.0);
+}
+
+TEST(LinearScoringFunctionTest, ReturnsNegativeValueForValueAboveThreshold) {
+  nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+  config.set_threshold(10.0);
+  config.set_k(1.0);
+  LinearScoringFunction scoring_function(config);
+  EXPECT_EQ(scoring_function.EvaluateMetric(12.0), -2.0);
+}
+
+TEST(SigmoidScoringFunctionTest, ReturnsZeroForValueEqualToThreshold) {
```

Comment:
For the sigmoid implementation especially, I wonder if it would make sense to generate a gold file with scipy's version of it, with a bunch of random numbers and some explicit edge cases as inputs, and verify that?
Also floating an idea -- I'm out of my league here -- but reading up [1] on the internets, I wonder if offering variants of it is worth something. Of course, that could be an excellent subject for experimentation in another plugin :-)

[1] https://stats.stackexchange.com/questions/330559/why-is-tanh-almost-always-better-than-sigmoid-as-an-activation-function

Reply:
Will come back to this in #433